---
title: "shot2_DGE"
author: "John Swenson"
date: "8/2/2021"
output: html_document
---

#### Chunk 1: Initial setup
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

rm(list=ls()) #Clear workspace

#Install bioconductor to install edgeR - run below two lines together
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("edgeR") #Install edgeR from Bioconductor
browseVignettes("edgeR") #Look up edgeR vigenette (if desired)

install.packages(c("tidyverse","statmod","RColorBrewer", "pheatmap", "ggpubr")) #Install other packages from CRAN
```

Here's a really useful resource: https://dockflow.org/workflow/rnaseq-gene-edgerql/


EdgeR DGE analysis
#### Chunk 2: Fit model and visualize data
```{r Fit model and visualize data}
library(edgeR)
library(tidyverse)
library(statmod)
library(RColorBrewer)
library(pheatmap)
library(ggpubr)

#Set working directory to location of counts matrix and gene_boundaries bed file: provide path between quotes
setwd("")

#-------------READ IN DATA------------------------
counts <- data.frame(data.matrix(read.table(file="all_counts_stranded.matrix"))) #Import count data to R

head(counts) #Make sure that samples are ordered the same as the vector in the line below (i.e. make sure the samples will be names appropriately)

#rename column names to sample names
colnames(counts) <- c("h1-4-H1_S10", "h1-4-H2_S12", "h1-4-H3_S3", "h1-4-RT1_S1", "h1-4-RT2_S6", "h1-4-RT3_S8","sh2-H1_S11", "sh2-H2_S2", "sh2-H3_S4", "sh2-RT1_S5", "sh2-RT2_S7", "sh2-RT3_S9")

#--------------DGE analysis by treatment---------------
#Group replicate samples
group <- factor(c(rep(c("h1.4_H", "h1.4_RT", "sh2_H", "sh2_RT"), each = 3))) 

y <- DGEList(counts = counts, group = group) #Create DGEList object with groups

#Filter out lowly expressed genes that will mess up comparisons
keep <- filterByExpr(y)
y <- y[keep,,keep.lib.sizes=FALSE]

#Set parameters and fit model
y <- calcNormFactors(y) #Normalize library sizes to account for some samples having genes that absorb a disproportionate amount of read depth
design <- model.matrix(~0+group) #Create design matrix (0+ replaces the intercept with a group)
y <- estimateDisp(y, design, robust=TRUE) #Estimate dispersion for negative binomial model

fit <- glmQLFit(y, design, robust=TRUE) #Fit the negative binomial model

#Make design matrix for all pairwise comparisons
comparisons <- makeContrasts(h1.4_H_vs_h1.4_RT = grouph1.4_H - grouph1.4_RT,
                             h1.4_H_vs_sh2_H = grouph1.4_H - groupsh2_H,
                             h1.4_H_vs_sh2_RT = grouph1.4_H - groupsh2_RT,
                             h1.4_RT_vs_sh2_H = grouph1.4_RT - groupsh2_H,
                             h1.4_RT_vs_sh2_RT = grouph1.4_RT - groupsh2_RT,
                             sh2_H_vs_sh2_RT = groupsh2_H - groupsh2_RT,
                             levels=design)

#Combine all comparisons into a single F statistic and p-value and examine top tags
comps <- glmQLFTest(fit, contrast=comparisons)
topTags(comps)

#------------Plots------------------
#Make MDS plot to ensure samples group together
points <- c(rep(c(19,18,15,16), each=3))
colors <- c(rep(c("blue", "red", "green", "purple"), each = 3))
plotMDS(y, col=colors, pch=points)
legend("top", legend=levels(group), pch=points[c(1,4,7,10)], col=colors[c(1,4,7,10)], ncol=1)

#Plot quasi-likelihood dispersion estimates - see edgeR manual for interpretation
plotQLDisp(fit)

##Make heatmap of all comparisons##
logcpm <- cpm(y, log=TRUE) #calculate counts per million
rownames(logcpm) <- rownames(y$counts) #save gene names as rownames
colnames(logcpm) <- paste(y$samples$group, sep="_") #name samples by group
o <- order(comps$table$PValue) #order differentially expressed genes by p value and save row index value
logcpm <- logcpm[o[1:30],] #extract counts per million of the highest DEGs specified above

#Create heat map
pheatmap(logcpm, cluster_col=FALSE, scale="row")
```


#### Chunk 3: Perform pairwise comparisons
```{r Perform DGE comparisons between treatments}
###Comparison 1: h1.4_H_vs_h1.4_RT
##Perform differential expression analysis
qlf_h1.4_H_vs_h1.4_RT <- glmQLFTest(fit, contrast=comparisons[,"h1.4_H_vs_h1.4_RT"]) 

#Show number of genes differentially expressed (up and down)
summary(decideTestsDGE(qlf_h1.4_H_vs_h1.4_RT)) 

#Extract all differentially expressed genes at FDR < 0.05 differentially expressed genes (DEGs) - set n to number of genes in dataset so all are included
top_DGEs1 <- topTags(qlf_h1.4_H_vs_h1.4_RT, n=17243)

#Save top DEGs in a csv file
write.csv(top_DGEs1, file="Top_tags_h1.4_H_vs_h1.4_RT_stranded.csv", row.names=TRUE)

##Plots
#Histogram of FDR values
top_DGEs1 <- as.data.frame(top_DGEs1)
hist(top_DGEs1$FDR, main="h1.4_H_vs_h1.4_RT", xlab = "FDR")
#Make MD plot of comparisons
plotMD(qlf_h1.4_H_vs_h1.4_RT, main = "h1.4_H_vs_h1.4_RT") 


###Comparison 2: h1.4_H_vs_sh2_H
##Perform differential expression analysis
qlf_h1.4_H_vs_sh2_H <- glmQLFTest(fit, contrast=comparisons[,"h1.4_H_vs_sh2_H"]) 

#Show number of genes differentially expressed (up and down)
summary(decideTestsDGE(qlf_h1.4_H_vs_sh2_H)) 

#Extract all differentially expressed genes at FDR < 0.05 differentially expressed genes (DEGs) - set n high so the limit is the p_value cutoff
top_DGEs2 <- topTags(qlf_h1.4_H_vs_sh2_H, n=17243)

#Save top DEGs in a csv file
write.csv(top_DGEs2, file="Top_tags_h1.4_H_vs_sh2_H_stranded.csv", row.names=TRUE)

##Plots
#Histogram of FDR values
top_DGEs2 <- as.data.frame(top_DGEs2)
hist(top_DGEs1$FDR, main="h1.4_H_vs_sh2_H", xlab = "FDR")
#Make MD plot of comparisons
plotMD(qlf_h1.4_H_vs_sh2_H, main = "h1.4_H_vs_sh2_H") 


###Comparison 3: h1.4_H_vs_sh2_RT
##Perform differential expression analysis
qlf_h1.4_H_vs_sh2_RT <- glmQLFTest(fit, contrast=comparisons[,"h1.4_H_vs_sh2_RT"]) #

#Show number of genes differentially expressed (up and down)
summary(decideTestsDGE(qlf_h1.4_H_vs_sh2_RT)) 

#Extract all differentially expressed genes at FDR < 0.05 differentially expressed genes (DEGs) - set n high so the limit is the p_value cutoff
top_DGEs3 <- topTags(qlf_h1.4_H_vs_sh2_RT, n=17243)

#Save top DEGs in a csv file
write.csv(top_DGEs3, file="Top_tags_h1.4_H_vs_sh2_RT_stranded.csv", row.names=TRUE)

##Plots
#Histogram of FDR values
top_DGEs3 <- as.data.frame(top_DGEs3)
hist(top_DGEs3$FDR, main="h1.4_H_vs_sh2_RT", xlab = "FDR")
#Make MD plot of comparisons
plotMD(qlf_h1.4_H_vs_sh2_RT, main = "h1.4_H_vs_sh2_RT") 


###Comparison 4: h1.4_RT_vs_sh2_H
##Perform differential expression analysis
qlf_h1.4_RT_vs_sh2_H <- glmQLFTest(fit, contrast=comparisons[,"h1.4_RT_vs_sh2_H"]) #

#Show number of genes differentially expressed (up and down)
summary(decideTestsDGE(qlf_h1.4_RT_vs_sh2_H)) 

#Extract all differentially expressed genes at FDR < 0.05 differentially expressed genes (DEGs) - set n high so the limit is the p_value cutoff
top_DGEs4 <- topTags(qlf_h1.4_RT_vs_sh2_H, n=17243)

#Save top DEGs in a csv file
write.csv(top_DGEs4, file="Top_tags_h1.4_RT_vs_sh2_H_stranded.csv", row.names=TRUE)

##Plots
#Histogram of FDR values
top_DGEs4 <- as.data.frame(top_DGEs1)
hist(top_DGEs4$FDR, main="h1.4_RT_vs_sh2_H", xlab = "FDR")
#Make MD plot of comparisons
plotMD(qlf_h1.4_RT_vs_sh2_H, main = "h1.4_RT_vs_sh2_H") 


###Comparison 5: h1.4_RT_vs_sh2_RT
##Perform differential expression analysis
qlf_h1.4_RT_vs_sh2_RT <- glmQLFTest(fit, contrast=comparisons[,"h1.4_RT_vs_sh2_RT"]) #

#Show number of genes differentially expressed (up and down)
summary(decideTestsDGE(qlf_h1.4_RT_vs_sh2_RT)) 

#Extract all differentially expressed genes at FDR < 0.05 differentially expressed genes (DEGs) - set n high so the limit is the p_value cutoff
top_DGEs5 <- topTags(qlf_h1.4_RT_vs_sh2_RT, n=17243)

#Save top DEGs in a csv file
write.csv(top_DGEs5, file="Top_tags_h1.4_RT_vs_sh2_RT_stranded.csv", row.names=TRUE)

##Plots
#Histogram of FDR values - need to include all samples above
top_DGEs5 <- as.data.frame(top_DGEs5)
hist(top_DGEs5$FDR, main="h1.4_RT_vs_sh2_RT", xlab = "FDR")
#Make MD plot of comparisons
plotMD(qlf_h1.4_RT_vs_sh2_RT, main = "h1.4_RT_vs_sh2_RT") 


###Comparison 6: sh2_H_vs_sh2_RT
##Perform differential expression analysis
qlf_sh2_H_vs_sh2_RT <- glmQLFTest(fit, contrast=comparisons[,"sh2_H_vs_sh2_RT"]) #

#Show number of genes differentially expressed (up and down)
summary(decideTestsDGE(qlf_sh2_H_vs_sh2_RT)) 

#Extract all differentially expressed genes at FDR < 0.05 differentially expressed genes (DEGs) - set n high so the limit is the p_value cutoff
top_DGEs6 <- topTags(qlf_sh2_H_vs_sh2_RT, n=17243)

#Save top DEGs in a csv file
write.csv(top_DGEs6, file="Top_tags_sh2_H_vs_sh2_RT_stranded.csv", row.names=TRUE)

##Plots
#Histogram of FDR values
top_DGEs6 <- as.data.frame(top_DGEs6)
hist(top_DGEs6$FDR, main="sh2_H_vs_sh2_RT", xlab = "FDR")
#Make MD plot of comparisons
plotMD(qlf_sh2_H_vs_sh2_RT, main = "sh2_H_vs_sh2_RT") 
```


#### Chunk 4: Change gene boundaries to detect readthrough
```{r Set gene boundaries that define "readthrough"}
#Assumes the count data from the first chunk have been read into R and lowly expressed genes filtered out

#----------------Set parameter values and read in data-----------------
#Set distance for upstream genic region and downstream readthrough region. Default  = 100bp
s <- 100


#Read in gene boundaries bed file for expressed genes (default) or all genes
genes <- read.table("gene_boundaries.bed") 
head(genes)

colnames(genes) <- c("chromosome", "start", "end","gene_name", "V5","strand", "database", "feature", "V9", "notes") #Rename columns
genes$gene_name <- as.character(genes$gene_name) #Convert columns to character

#Subset all genes for expressed genes
expressed_genes <- rownames_to_column(as.data.frame(y$counts), var="gene_name")  %>% #subset for genes that pass edgeR's filters
  as_tibble() %>% #convert to tibble  
  inner_join(genes, by = "gene_name") %>% #join gene boundaries bed file and table of expressed genes, keeping only expressed genes
  select(chromosome, start, end, gene_name, V5, strand, database, feature, V9, notes) %>% #select relevant columns
  filter(chromosome != "ChrC" & chromosome != "ChrM") #Remove chloroplast and mitochondrial sequences

head(expressed_genes)

nrow(genes) #How many genes are present in the full bed file that inlcudes all genes?
nrow(expressed_genes) #How many genes are included in the bed file that's been filtered for expressed genes?
genes$gene_name[!genes$gene_name %in% expressed_genes$gene_name] #Which genes are in the annotation file and NOT expressed?

#Write file including full gene boundaries to use for intersecting with bedtools
#(expressed_genes, file = "expressed_gene_boundaries_stranded.bed", sep="\t", col.names = FALSE, row.names = FALSE, quote = FALSE)

# --------Create boundaries for readthrough detection--------

#Make bed file for genic region.
##Normalize by taking the region s base pairs upstream of the transcription stop site for each gene. 
#If the strand is negative, then set the end to the start (b/c it goes the other way) and set start to s base pairs upstream
expressed_genes_upstream_region <- expressed_genes %>% 
  mutate(start2 = ifelse(strand == "-", start, end - (s-1))) %>% 
  mutate(end2 = ifelse(strand == "-", start + (s-1), end)) %>% 
  select(chromosome, start2, end2, gene_name, V5, strand, database, feature, V9, notes)
head(expressed_genes_upstream_region)

write.table(expressed_genes_upstream_region, file = "expressed_genes_upstream_region.bed", sep="\t", col.names = FALSE, row.names = FALSE, quote = FALSE)

#Make bed file for "readthrough" i.e. downstream region
#Use window of 100bp, and similarly switch start and stop for genes on the negative strand (b/c they go the other way)
expressed_genes_downstream_region <- expressed_genes %>% 
  mutate(start2 = ifelse(strand == "-", start - s, end + 1)) %>% 
  mutate(end2 = ifelse(strand == "-", start - 1, end + s)) %>% 
  select(chromosome, start2, end2, gene_name, V5, strand, database, feature, V9, notes)
head(expressed_genes_downstream_region)  
  
write.table(expressed_genes_downstream_region, file = "expressed_genes_downstream_region.bed", sep="\t", col.names = FALSE, row.names = FALSE, quote = FALSE)
```